File: examples/csharp/avalonia/NotesShowcase/ViewModels/StatusBarVM.cs
using System.Reactive.Linq;
using System.Reactive.Subjects;
using VMx.Builders;
using VMx.Components;
using VMx.Messages;
using VMx.Properties;
using VMx.Services;
using NotesShowcase.Views.Adapter;

namespace NotesShowcase.ViewModels;

/// <summary>
/// Read-only VM driving the three status-bar slots: note count, starred count,
/// and editing state. Each slot is a <see cref="DerivedProperty{TValue}"/>
/// composed from hub-published <see cref="PropertyChangedMessage{TSender}"/>
/// streams.
/// </summary>
public sealed class StatusBarVM : ComponentVMBase
{
    private readonly NotesViewVM _notesView;
    private readonly NoteFormVM _noteForm;
    private readonly BehaviorSubject<NotesViewVM> _notesViewSubject;
    private readonly BehaviorSubject<NoteFormVM> _noteFormSubject;
    private readonly IDisposable _notesViewSub;
    private readonly IDisposable _noteFormSub;
    private bool _ownDisposed;

    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Component;

    /// <summary>Public hub accessor.</summary>
    public new IMessageHub Hub => base.Hub;

    /// <summary>"N notes" — total count across all visible items in <see cref="NotesViewVM"/>.</summary>
    public DerivedProperty<string> NoteCountText { get; }

    /// <summary>"K starred" — count of starred items.</summary>
    public DerivedProperty<string> StarredText { get; }

    /// <summary>"Editing: TITLE" / "No selection" — reflects the form's draft state.</summary>
    public DerivedProperty<string> EditingText { get; }

    // Phase 5.a binding gap #3 (parity with Phase 5.b ``bind_derived_property``
    // and Phase 5.c ``useDerivedProperty``): DerivedProperty does NOT raise
    // INotifyPropertyChanged for its ``.Value`` getter, so Avalonia's binding
    // engine cannot observe recomputes through ``{Binding Foo.Value}``. The
    // ``Bindable*`` parallels below wrap each DP with a ``BindableDerived<T>``
    // sidecar that raises ``PropertyChanged("Value")`` on every distinct
    // emission. Views bind ``{Binding NoteCountTextBindable.Value}`` instead.

    /// <summary>INPC-aware sidecar for <see cref="NoteCountText"/>.</summary>
    public BindableDerived<string> NoteCountTextBindable { get; }

    /// <summary>INPC-aware sidecar for <see cref="StarredText"/>.</summary>
    public BindableDerived<string> StarredTextBindable { get; }

    /// <summary>INPC-aware sidecar for <see cref="EditingText"/>.</summary>
    public BindableDerived<string> EditingTextBindable { get; }

    private StatusBarVM(
        string name,
        string hint,
        IMessageHub hub,
        IDispatcher dispatcher,
        NotesViewVM notesView,
        NotebooksRootVM notebooks,
        NoteFormVM noteForm)
        : base(name, hint, hub, dispatcher, onConstruct: null, onDestruct: null)
    {
        _notesView = notesView;
        _ = notebooks; // builder surface kept for cross-flavor parity
        _noteForm = noteForm;

        _notesViewSubject = new BehaviorSubject<NotesViewVM>(notesView);
        _noteFormSubject = new BehaviorSubject<NoteFormVM>(noteForm);

        // Re-emit each source whenever its hub publishes a relevant PropertyChanged.
        // live binding: hub messages can arrive from background
        // continuations; the subjects feed BindableDerived, which raises
        // INPC on the caller's thread — marshal to the foreground.
        _notesViewSub = notesView.Hub.Messages
            .OfType<PropertyChangedMessage<IComponentVM>>()
            .Where(m => ReferenceEquals(m.Sender, notesView))
            .ObserveOn(dispatcher.Foreground)
            .Subscribe(_ => _notesViewSubject.OnNext(notesView));
        _noteFormSub = noteForm.Hub.Messages
            .OfType<PropertyChangedMessage<IComponentVM>>()
            .Where(m => ReferenceEquals(m.Sender, noteForm))
            .ObserveOn(dispatcher.Foreground)
            .Subscribe(_ => _noteFormSubject.OnNext(noteForm));

        NoteCountText = DerivedProperty.From(
            _notesViewSubject,
            nv => $"{nv.FilteredItems.Count} note{(nv.FilteredItems.Count == 1 ? string.Empty : "s")}");

        StarredText = DerivedProperty.From(
            _notesViewSubject,
            nv =>
            {
                var k = nv.FilteredItems.Count(n => n.Starred);
                return $"{k} starred";
            });

        EditingText = DerivedProperty.From(
            _noteFormSubject,
            nf => nf.HasBoundNote
                ? $"Editing: {nf.Draft.Title}{(nf.IsDirty ? " *" : string.Empty)}"
                : "No selection");

        NoteCountTextBindable = new BindableDerived<string>(NoteCountText);
        StarredTextBindable = new BindableDerived<string>(StarredText);
        EditingTextBindable = new BindableDerived<string>(EditingText);
    }

    /// <inheritdoc/>
    public override void Dispose()
    {
        if (_ownDisposed) { base.Dispose(); return; }
        _ownDisposed = true;
        NoteCountTextBindable.Dispose();
        StarredTextBindable.Dispose();
        EditingTextBindable.Dispose();
        NoteCountText.Dispose();
        StarredText.Dispose();
        EditingText.Dispose();
        _notesViewSub.Dispose();
        _noteFormSub.Dispose();
        _notesViewSubject.OnCompleted(); _notesViewSubject.Dispose();
        _noteFormSubject.OnCompleted(); _noteFormSubject.Dispose();
        base.Dispose();
    }

    /// <summary>Returns a new empty builder.</summary>
    public static StatusBarVMBuilder Builder() => StatusBarVMBuilder.Empty;

    /// <summary>Immutable fluent builder.</summary>
    public sealed class StatusBarVMBuilder
    {
        private readonly string? _name;
        private readonly string _hint;
        private readonly IMessageHub? _hub;
        private readonly IDispatcher? _dispatcher;
        private readonly NotesViewVM? _notesView;
        private readonly NotebooksRootVM? _notebooks;
        private readonly NoteFormVM? _noteForm;

        internal static readonly StatusBarVMBuilder Empty = new();
        private StatusBarVMBuilder() { _hint = ""; }
        private StatusBarVMBuilder(
            string? name, string hint,
            IMessageHub? hub, IDispatcher? dispatcher,
            NotesViewVM? notesView, NotebooksRootVM? notebooks, NoteFormVM? noteForm)
        {
            _name = name; _hint = hint;
            _hub = hub; _dispatcher = dispatcher;
            _notesView = notesView; _notebooks = notebooks; _noteForm = noteForm;
        }
        /// <summary>Sets the required Name.</summary>
        public StatusBarVMBuilder Name(string name) => new(name, _hint, _hub, _dispatcher, _notesView, _notebooks, _noteForm);
        /// <summary>Sets the optional Hint.</summary>
        public StatusBarVMBuilder Hint(string hint) => new(_name, hint, _hub, _dispatcher, _notesView, _notebooks, _noteForm);
        /// <summary>Sets the required Services.</summary>
        public StatusBarVMBuilder Services(IMessageHub hub, IDispatcher dispatcher) => new(_name, _hint, hub, dispatcher, _notesView, _notebooks, _noteForm);
        /// <summary>Sets the required NotesViewVM source.</summary>
        public StatusBarVMBuilder NotesView(NotesViewVM source) => new(_name, _hint, _hub, _dispatcher, source, _notebooks, _noteForm);
        /// <summary>Sets the required NotebooksRootVM source.</summary>
        public StatusBarVMBuilder Notebooks(NotebooksRootVM source) => new(_name, _hint, _hub, _dispatcher, _notesView, source, _noteForm);
        /// <summary>Sets the required NoteFormVM source.</summary>
        public StatusBarVMBuilder NoteForm(NoteFormVM source) => new(_name, _hint, _hub, _dispatcher, _notesView, _notebooks, source);

        /// <summary>Builds the VM after validation.</summary>
        public StatusBarVM Build()
        {
            BuilderValidationException.Require(_name, "Name");
            BuilderValidationException.Require(_hub, "Hub");
            BuilderValidationException.Require(_dispatcher, "Dispatcher");
            BuilderValidationException.Require(_notesView, "NotesView");
            BuilderValidationException.Require(_notebooks, "Notebooks");
            BuilderValidationException.Require(_noteForm, "NoteForm");
            return new StatusBarVM(_name!, _hint, _hub!, _dispatcher!, _notesView!, _notebooks!, _noteForm!);
        }
    }
}

File: examples/typescript/react/notes-showcase/src/viewmodels/statusBarVM.ts
/**
 * StatusBarVM — read-only VM driving the three status-bar slots:
 *   * `noteCountText` — "N notes" / "1 note"
 *   * `starredText`   — "K starred"
 *   * `editingText`   — "Editing: TITLE" / "Editing: TITLE *" / "No selection"
 *
 * Each slot is a `DerivedProperty<string>`, recomputed from hub-published
 * PropertyChangedMessages from the upstream `NotesViewVM` / `NoteFormVM`.
 * Equality-guarded internally by DerivedProperty (Object.is on the new value).
 */
import { BehaviorSubject, filter, map, merge, type Subscription } from "rxjs";
import {
  ComponentVMBase,
  DerivedProperty,
  PropertyChangedMessage,
  ViewModelType,
  type IDispatcher,
  type IMessageHub,
} from "@thekaveh/vmx";

import type { NoteFormVM } from "./noteFormVM.js";
import type { NotebooksRootVM } from "./notebooksRootVM.js";
import type { NotesViewVM } from "./notesViewVM.js";

export class StatusBarVM extends ComponentVMBase {
  readonly #notesView: NotesViewVM;
  readonly #noteForm: NoteFormVM;
  readonly #notebooks: NotebooksRootVM;
  readonly #notesViewTick: BehaviorSubject<void>;
  readonly #noteFormTick: BehaviorSubject<void>;
  readonly #hubSubs: Subscription[] = [];
  readonly #noteCountText: DerivedProperty<string>;
  readonly #starredText: DerivedProperty<string>;
  readonly #editingText: DerivedProperty<string>;

  constructor(opts: {
    name: string;
    hint: string;
    hub: IMessageHub;
    dispatcher: IDispatcher;
    notesView: NotesViewVM;
    notebooks: NotebooksRootVM;
    noteForm: NoteFormVM;
  }) {
    super({
      name: opts.name,
      hint: opts.hint,
      hub: opts.hub,
      dispatcher: opts.dispatcher,
    });
    this.#notesView = opts.notesView;
    this.#notebooks = opts.notebooks;
    this.#noteForm = opts.noteForm;

    this.#notesViewTick = new BehaviorSubject<void>(undefined);
    this.#noteFormTick = new BehaviorSubject<void>(undefined);

    const notesViewMsgs = opts.notesView.hub.messages.pipe(
      filter(
        (m): m is PropertyChangedMessage<unknown> =>
          m instanceof PropertyChangedMessage && m.sender === opts.notesView,
      ),
    );
    this.#hubSubs.push(notesViewMsgs.subscribe(() => this.#notesViewTick.next()));

    const noteFormMsgs = opts.noteForm.hub.messages.pipe(
      filter(
        (m): m is PropertyChangedMessage<unknown> =>
          m instanceof PropertyChangedMessage && m.sender === opts.noteForm,
      ),
    );
    this.#hubSubs.push(noteFormMsgs.subscribe(() => this.#noteFormTick.next()));

    const notesViewStream = merge(this.#notesViewTick).pipe(
      map(() => {
        const n = this.#notesView.filteredItems.length;
        const k = this.#notesView.filteredItems.filter((x) => x.model.starred)
          .length;
        return { n, k };
      }),
    );

    const formStream = merge(this.#noteFormTick).pipe(
      map(() => ({
        bound: this.#noteForm.hasBoundNote,
        title: this.#noteForm.draft.title,
        dirty: this.#noteForm.isDirty,
      })),
    );

    this.#noteCountText = new DerivedProperty<string>(
      notesViewStream.pipe(
        map(({ n }) => `${n} note${n === 1 ? "" : "s"}`),
      ),
      null,
      null,
    );
    this.#starredText = new DerivedProperty<string>(
      notesViewStream.pipe(map(({ k }) => `${k} starred`)),
      null,
      null,
    );
    this.#editingText = new DerivedProperty<string>(
      formStream.pipe(
        map(({ bound, title, dirty }) =>
          bound
            ? `Editing: ${title}${dirty ? " *" : ""}`
            : "No selection",
        ),
      ),
      null,
      null,
    );
  }

  get type(): ViewModelType {
    return ViewModelType.Component;
  }

  get hub(): IMessageHub {
    return this._hub;
  }

  get notebooks(): NotebooksRootVM {
    return this.#notebooks;
  }

  get noteCountText(): DerivedProperty<string> {
    return this.#noteCountText;
  }

  get starredText(): DerivedProperty<string> {
    return this.#starredText;
  }

  get editingText(): DerivedProperty<string> {
    return this.#editingText;
  }

  protected override _onDispose(): void {
    // The hub outlives this VM — drop our callbacks so a disposed status bar
    // stops ticking (the Python flagship already does this via self._subs).
    for (const sub of this.#hubSubs) sub.unsubscribe();
    this.#noteCountText.dispose();
    this.#starredText.dispose();
    this.#editingText.dispose();
    this.#notesViewTick.complete();
    this.#noteFormTick.complete();
    super._onDispose();
  }

  static builder(): StatusBarVMBuilder {
    return new StatusBarVMBuilder();
  }
}

export class StatusBarVMBuilder {
  #name: string | null = null;
  #hint = "";
  #hub: IMessageHub | null = null;
  #dispatcher: IDispatcher | null = null;
  #notesView: NotesViewVM | null = null;
  #notebooks: NotebooksRootVM | null = null;
  #noteForm: NoteFormVM | null = null;

  constructor(from?: StatusBarVMBuilder) {
    if (from) {
      this.#name = from.#name;
      this.#hint = from.#hint;
      this.#hub = from.#hub;
      this.#dispatcher = from.#dispatcher;
      this.#notesView = from.#notesView;
      this.#notebooks = from.#notebooks;
      this.#noteForm = from.#noteForm;
    }
  }

  name(value: string): StatusBarVMBuilder {
    const b = new StatusBarVMBuilder(this);
    b.#name = value;
    return b;
  }

  hint(value: string): StatusBarVMBuilder {
    const b = new StatusBarVMBuilder(this);
    b.#hint = value;
    return b;
  }

  services(hub: IMessageHub, dispatcher: IDispatcher): StatusBarVMBuilder {
    const b = new StatusBarVMBuilder(this);
    b.#hub = hub;
    b.#dispatcher = dispatcher;
    return b;
  }

  notesView(vm: NotesViewVM): StatusBarVMBuilder {
    const b = new StatusBarVMBuilder(this);
    b.#notesView = vm;
    return b;
  }

  notebooks(vm: NotebooksRootVM): StatusBarVMBuilder {
    const b = new StatusBarVMBuilder(this);
    b.#notebooks = vm;
    return b;
  }

  noteForm(vm: NoteFormVM): StatusBarVMBuilder {
    const b = new StatusBarVMBuilder(this);
    b.#noteForm = vm;
    return b;
  }

  build(): StatusBarVM {
    if (this.#name === null) throw new Error("name is required");
    if (this.#hub === null || this.#dispatcher === null)
      throw new Error("services are required");
    if (this.#notesView === null) throw new Error("notesView is required");
    if (this.#notebooks === null) throw new Error("notebooks is required");
    if (this.#noteForm === null) throw new Error("noteForm is required");
    return new StatusBarVM({
      name: this.#name,
      hint: this.#hint,
      hub: this.#hub,
      dispatcher: this.#dispatcher,
      notesView: this.#notesView,
      notebooks: this.#notebooks,
      noteForm: this.#noteForm,
    });
  }
}
